_posts/2016-05-22-th1eadin0g1.markdown

```diff
@@ -0,0 +1,202 @@
+
+This could be as simple as downloading a file, or image from the internet, accessing an external API, using something like background locations or playing sounds. There are no limits to the amount of cases where you could be doing some asynchronous work under the hood. So, I'm going to try and cover _most_ of the common use cases of where you are using it without knowing, then expand on a few techniques to write useful multi-threaded code.
+
+<!-- more -->
+
+### Thread Terminology 101
+
+* `Process` - The context within an app is running, you will have one process per app. 
+* `Thread` - A pathway of code, you can have multiple threads in an app's process.
+* `Main Thread` - The main thread for your process, this is the only thread you can do user interface code.
+* `Blocked` - You're waiting for some work to happen on a thread, so that something else can happen.
+* `Deadlock` - One thread is waiting for another thread to finish, but that thread is waiting for the other. Neither finish.   
+
+### Thinking POSIX-itive
+
+Let's just say this is elephants all the way down, but we should at least know the terminology so you can pretend to talk like you know the buzzwords. 
+
+So, pretty much all threading implementations are based on [POSIX threads](https://en.wikipedia.org/wiki/POSIX_Threads). What is POSIX you say? Well POSIX is a standard for communications inside an operating system. You would say an Operating System (OS) is POSIX compliant in that it conforms to the standard, iOS/OS X and Linux do, Android and Windows "kinda" do. It's more like a "It's complicated" relation. You don't need to know the details.
+
+A POSIX thread is commonly called a `pthread`. The POSIX standard for `pthread`s is a collection of c functions. So they look a bit like `pthread_atfork`, `pthread_attr_destroy`, `pthread_attr_getdetachstate`, `pthread_attr_getguardsize` etc etc. You can open an Xcode project and do "pthreads.h" in "Open Quickly.." to find them all.
+
+These provide the lowest level API to threads, everything is built on top of this.
+
+### "Same Time"
+
+Saying that something is happening at the same time, is potentially a lie, or more accurately a simplification. I'm currently writing this on a MacBook Pro, with these specs: 
+
+```
+Hardware Overview:
+
+  Model Name:	MacBook Pro
+  Model Identifier:	MacBookPro12,1
+  Processor Name:	Intel Core i7
+  Processor Speed:	3.1 GHz
+  Number of Processors:	1
+  Total Number of Cores:	2
+  L2 Cache (per Core):	256 KB
+  L3 Cache:	4 MB
+  Memory:	16 GB
+```
+
+Which means there's one processor with two cores. As a rule of thumb, a processor can only process one instruction at a time. However, a processor can fake being more than one processor at once. Which is essentially what cores mean. 
```

Comment:
You’re describing hyper-threading. Actual cores are independent processing units, they can perform multiple operations at the same time, aka true parallel computing.

* https://en.wikipedia.org/wiki/Hyper-threading
* https://en.wikipedia.org/wiki/Multi-core_processor